File: addtwonumbers.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	result := &ListNode{}
	head := result

	cur1 := l1
	cur2 := l2

	var carry *int

	for cur1 != nil && cur2 != nil {
		sum := cur1.Val + cur2.Val

		if sum > 10 {
			c := sum / 10
			carry = &c
			result.Val = sum % 10

		} else {

		}

		cur1 = cur1.Next
		cur2 = cur2.Next

		result.Next = &ListNode{}
		result = result.Next
	}

	return head
}

func main() {
	l1 := &ListNode{Val: 2, Next: &ListNode{Val: 4, Next: &ListNode{Val: 3}}}
	l2 := &ListNode{Val: 5, Next: &ListNode{Val: 6, Next: &ListNode{Val: 4}}}

	sum := addTwoNumbers(l1, l2)

	for sum.Next != nil {
		fmt.Println(sum.Val)
		sum = sum.Next
	}
}

File: add_two_numbers.py
from typing import List, Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        digits = None
        head = digits
        carry = 0

        while l1 != None and l2 != None:
            sum = l1.val + l2.val + carry
            carry = 0

            if sum >= 10:
                result = sum % 10
                carry = int(sum / 10)
            else:
                result = sum

            if digits == None:
                digits = ListNode(val=result)
                head = digits
            else:
                digits.next = ListNode(val=result)
                digits = digits.next

            l1 = l1.next
            l2 = l2.next

        while l1 != None:
            sum = l1.val + carry
            carry = 0

            if sum >= 10:
                result = sum % 10
                carry = int(sum / 10)
            else:
                result = sum

            if digits == None:
                digits = ListNode(val=result)
                head = digits
            else:
                digits.next = ListNode(val=result)
                digits = digits.next

            l1 = l1.next

        while l2 != None:
            sum = l2.val + carry
            carry = 0

            if sum >= 10:
                result = sum % 10
                carry = int(sum / 10)
            else:
                result = sum

            if digits == None:
                digits = ListNode(val=result)
                head = digits
            else:
                digits.next = ListNode(val=result)
                digits = digits.next

            l2 = l2.next

        if carry > 0:
            digits.next = ListNode(val=carry)

        return head


if __name__ == "__main__":
    # l1 = ListNode(2, ListNode(4, ListNode(3)))
    # l1 = ListNode(2, ListNode(4, ListNode(5)))
    # l1 = ListNode(9, ListNode(9, ListNode(9)))
    # l2 = ListNode(1)
    l1 = ListNode(1)
    l2 = ListNode(9, ListNode(9, ListNode(9)))
    # l2 = ListNode(5, ListNode(6, ListNode(4)))

    s = Solution()
    digits = s.addTwoNumbers(l1, l2)

    while digits != None:
        print(digits.val)
        digits = digits.next
